app/models/concerns/user_concern.rb

```diff
@@ -151,6 +156,34 @@ module UserConcern
       where(permission_context: [nil, 'role_based'])
     end
 
+    # datetime of most recent warehouse login
+    def last_warehouse_login
+      # for non-HMIS installations, we can use the devise current_sign_in_at directly because there is only one scope
+      return current_sign_in_at unless HmisEnforcement.hmis_enabled?
+
+      # Find most recent Warehouse login using cached map of LoginActivities
+      LoginActivity.latest_warehouse_logins[id]
+    end
+
+    # datetime of most recent HMIS login
+    def last_hmis_login
+      return unless HmisEnforcement.hmis_enabled?
+
+      # Find most recent HMIS login using cached map of LoginActivities
+      LoginActivity.latest_hmis_logins[id]
+    end
+
+    # whether the user's most recent login was to the Warehouse or HMIS
+    def last_login_application
+      if last_warehouse_login && last_hmis_login
+        last_warehouse_login > last_hmis_login ? 'Warehouse' : 'HMIS'
+      elsif last_warehouse_login
+        'Warehouse'
+      elsif last_hmis_login
+        'HMIS'
+      end
```

Comment:
What about something like:
```suggestion
      [['Warehouse', last_warehouse_login], ['HMIS', last_hmis_login]].select { |_, date| date.present? }&.max_by(&:last)&.first
```

There's probably something slightly more descriptive, but this would decrease some of the duplication.